### PR TITLE
[BITAU-149] Add Setting to Turn On Authenticator Syncing for an Account

### PR DESCRIPTION
**Pull request by brant-livefront**
## 🎟️ Tracking

[[BITAU-149](https://livefront.atlassian.net/browse/BITAU-149)]

## 📔 Objective

This PR adds a `syncToAuthenticator` setting that will allow the sync to be turned on/off for specific accounts in the BW PM app. It was very closely modeled after all of the `connectToWatch` pieces. There is no UI to turn this on (that will be added last), but putting this in now allows us to begin hooking off the setting when doing the sync. 

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: BitwardenShared/Core/Platform/Services/StateServiceTests.swift [hunk 1]
```diff
@@ -1773,6 +1789,64 @@ class StateServiceTests: BitwardenTestCase { // swiftlint:disable:this type_body
...
+    }
+
+    /// `syncToAuthenticatorPublisher()` returns false if the user is not logged in.
+    func test_syncToAuthenticatorPublisher_notLoggedIn() async throws {
+        var publishedValues = [SyncToAuthenticatorValue]()
+        let publisher = await subject.syncToAuthenticatorPublisher()
+            .sink(receiveValue: { userId, shouldSync in
+                publishedValues.append(SyncToAuthenticatorValue(userId: userId, shouldSync: shouldSync))
+            })
+        defer { publisher.cancel() }
```
Comment (victor-livefront) [inline BitwardenShared/Core/Platform/Services/StateServiceTests.swift:1845]: These defers are nice. I've been putting them at the end for publishers and Tasks in tests, but I'm going to change to defers :)

**Review by victor-livefront — APPROVED** [review]
This looks good to me!

File: BitwardenShared/Core/Platform/Services/StateService.swift [hunk 2]
```diff
@@ -1713,6 +1765,20 @@ actor DefaultStateService: StateService { // swiftlint:disable:this type_body_le
...
 
+    func syncToAuthenticatorPublisher() async -> AnyPublisher<(String?, Bool), Never> {
+        activeAccountIdPublisher().flatMap { userId in
+            self.syncToAuthenticatorByUserIdSubject.map { values in
+                let userValue = if let userId {
+                    values[userId] ?? self.appSettingsStore.syncToAuthenticator(userId: userId)
+                } else {
+                    false
+                }
+                return (userId, userValue)
```
Comment (fedemkr) [inline BitwardenShared/Core/Platform/Services/StateService.swift:1776]: 🎨 Could this be rewritten as the following? IMO it's easier to read and follow the logic.
```suggestion
                guard let userId else {
                    return (nil, false)
                }

                let userValue = values[userId] ?? self.appSettingsStore.syncToAuthenticator(userId: userId)
                return (userId, userValue)
```

File: BitwardenShared/Core/Platform/Services/StateServiceTests.swift [hunk 3]
```diff
@@ -1862,4 +1936,9 @@ class StateServiceTests: BitwardenTestCase { // swiftlint:disable:this type_body
 private struct ConnectToWatchValue: Equatable {
     let userId: String?
     let shouldConnect: Bool
+}
+
+private struct SyncToAuthenticatorValue: Equatable {
+    let userId: String?
+    let shouldSync: Bool
 } // swiftlint:disable:this file_length
```
Comment (fedemkr) [inline BitwardenShared/Core/Platform/Services/StateServiceTests.swift:1944]: 🤔 Is this struct really necessary? Can't we use the published tuple directly?

File: BitwardenShared/Core/Platform/Services/StateService.swift (hunk 2 above)
Comment (brant-livefront) [reply BitwardenShared/Core/Platform/Services/StateService.swift]: I like this a lot better. 👍  Adopted in latest commit.

File: BitwardenShared/Core/Platform/Services/StateServiceTests.swift (hunk 3 above)
Comment (brant-livefront) [reply BitwardenShared/Core/Platform/Services/StateServiceTests.swift]: It's not totally necessary but you get Equatability doing it that way which cleans up the XCTAsserts a bit:

with the struct:
``` Swift
        XCTAssertEqual(
            publishedValues,
            [
                SyncToAuthenticatorValue(userId: "1", shouldSync: false),
                SyncToAuthenticatorValue(userId: "1", shouldSync: true),
            ]
        )
```

without the struct:
``` Swift
        XCTAssertEqual(publishedValues[0].userId, "1")
        XCTAssertEqual(publishedValues[0].shouldSync, false)
        XCTAssertEqual(publishedValues[1].userId, "1")
        XCTAssertEqual(publishedValues[1].shouldSync, true)
```

The latest commit has it without the struct. See which one you prefer - I'm fine either way 👍

File: BitwardenShared/Core/Platform/Services/StateServiceTests.swift [hunk 4]
```diff
@@ -1817,34 +1814,32 @@ class StateServiceTests: BitwardenTestCase { // swiftlint:disable:this type_body
...
-            publishedValues,
-            [
-                SyncToAuthenticatorValue(userId: "1", shouldSync: true),
-                SyncToAuthenticatorValue(userId: "1", shouldSync: false),
-            ]
-        )
+        XCTAssertEqual(publishedValues[0].userId, "1")
+        XCTAssertEqual(publishedValues[0].shouldSync, true)
+        XCTAssertEqual(publishedValues[1].userId, "1")
+        XCTAssertEqual(publishedValues[1].shouldSync, false)
```
Comment (fedemkr) [inline BitwardenShared/Core/Platform/Services/StateServiceTests.swift:1829]: 🎨 You can use `==` operator to reduce this a bit more:
```suggestion
        XCTAssertTrue(publishedValues[0] == (userId: "1", shouldSync: false))
        XCTAssertTrue(publishedValues[1] == (userId: "1", shouldSync: true))
```

File: BitwardenShared/Core/Platform/Services/StateServiceTests.swift (hunk 3 above)
Comment (fedemkr) [reply BitwardenShared/Core/Platform/Services/StateServiceTests.swift]: Yeah, it's a pity we can't use `Equatable` with tuples, but we can use `==` operator 😄

File: BitwardenShared/Core/Platform/Services/StateServiceTests.swift (hunk 4 above)
Comment (brant-livefront) [reply BitwardenShared/Core/Platform/Services/StateServiceTests.swift]: Ah, good call. Update all of them to this style in latest. 👍

Comment (brant-livefront) [reply BitwardenShared/Core/Platform/Services/StateServiceTests.swift]: Sorry, didn't see @KatherineInCode's comment when I updated. Let me know which one y'all prefer. I'm good either way.

**Review by victor-livefront — APPROVED** [review]
Updates look good to me.

File: BitwardenShared/Core/Platform/Services/StateServiceTests.swift (hunk 4 above)
Comment (KatherineInCode) [reply BitwardenShared/Core/Platform/Services/StateServiceTests.swift]: @brant-livefront Let's go with the `XCTAssertEquals` approach. Depending on how often we need to check the tuple in tests, [a custom assertion function](https://qualitycoding.org/swift-test-assertions/) might also be helpful.

Comment (brant-livefront) [reply BitwardenShared/Core/Platform/Services/StateServiceTests.swift]: @KatherineInCode Sounds good, I'll roll back to using explicit `XCTAssertEquals`.

Comment (brant-livefront) [reply BitwardenShared/Core/Platform/Services/StateServiceTests.swift:1829]: @KatherineInCode this is updated in the latest commit. 👍

**Review by KatherineInCode — APPROVED** [review]

**Pull request merged**